### DISCUSSION TITLE
Exception: An existing connection was forcibly closed by the remote host

### DISCUSSION THREAD

**Discussion by furqansafdar** [CoreWCF/CoreWCF#1283] [Q&A]
I am facing a situation where getting an exception when trying to make a service call from the client and this exception is coming on client's end. However, the server is successfully processing the request and returning the response.

> System.ServiceModel.CommunicationException
>   HResult=0x80131500
>   Message=An error occurred while sending the request.
>   Source=System.Private.ServiceModel
>   StackTrace:
>    at System.Runtime.AsyncResult.End[TAsyncResult](IAsyncResult result)
>    at System.ServiceModel.Channels.ServiceChannel.SendAsyncResult.End(SendAsyncResult result)
>    at System.ServiceModel.Channels.ServiceChannel.EndCall(String action, Object[] outs, IAsyncResult result)
>    at System.ServiceModel.Channels.ServiceChannelProxy.TaskCreator.<>c__DisplayClass1_0.<CreateGenericTask>b__0(IAsyncResult asyncResult)
> 
> Inner Exception 1:
> HttpRequestException: An error occurred while sending the request.
> 
> Inner Exception 2:
> IOException: Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host..
> 
> Inner Exception 3:
> SocketException: An existing connection was forcibly closed by the remote host.

```csharp
var builder = WebApplication.CreateBuilder();

builder.Services.AddServiceModelServices();
builder.Services.AddServiceModelMetadata();
builder.Services.AddSingleton<IServiceBehavior, UseRequestHeadersForMetadataAddressBehavior>();

var app = builder.Build();

app.UseServiceModel(serviceBuilder =>
{
    serviceBuilder.AddService<MyServiceManager>();
    serviceBuilder.AddServiceEndpoint<MyServiceManager, IMyService>(new BasicHttpBinding(BasicHttpSecurityMode.None), "/MyService.svc");

    var serviceMetadataBehavior = app.Services.GetRequiredService<ServiceMetadataBehavior>();
    serviceMetadataBehavior.HttpsGetEnabled = true;
});

app.Run();
```

At first, I configured the service with BasicHttpSecurityMode.Transport but then I changed it to BasicHttpSecurityMode.None just to rule out this issue. I even turned off my firewall on my local machine (based on the research on the web for this issue) and used fiddler to inspect the traffic and find out that it was giving a 504 error.

But none of the solutions really worked in my case.

Reaching out to community for suggestion to fix my issue. Please note I am using .NET 6.

**Comment by furqansafdar** [comment]
After continuing to investigate this issue, I have found that this is occuring due to circular dependency in the object graph.

> "Object graph for type 'Domain.SetValue' contains cycles and cannot be serialized if references are not tracked. Consider using the DataContractAttribute with the IsReference property set to true."

How can I solve it using without setting IsReference property to true?

I remember when I faced the similar issue in .NET Framework, I used the following code and applied the Attribute to my Service Contract and it was all working fine there but this same technique is not working in CoreWCF, any suggestions?

```csharp

[ServiceContract]
public interface IMyService
{
    [OperationContract]
    [ReferencePreservingDataContractFormat]
    CatalogResponse? SaveApplication(CatalogRequest catalogRequest);
}

public class ReferencePreservingDataContractFormatAttribute : Attribute, IOperationBehavior
{
    #region IOperationBehavior Members

    public void AddBindingParameters(OperationDescription description, BindingParameterCollection parameters)
    {
    }

    public void ApplyClientBehavior(OperationDescription description, ClientOperation proxy)
    {
        IOperationBehavior innerBehavior = new ReferencePreservingDataContractSerializerOperationBehavior(description);
        innerBehavior.ApplyClientBehavior(description, proxy);
    }

    public void ApplyDispatchBehavior(OperationDescription description, DispatchOperation dispatch)
    {
        IOperationBehavior innerBehavior = new ReferencePreservingDataContractSerializerOperationBehavior(description);
        innerBehavior.ApplyDispatchBehavior(description, dispatch);
    }

    public void Validate(OperationDescription description)
    {
    }

    #endregion
}

class ReferencePreservingDataContractSerializerOperationBehavior : DataContractSerializerOperationBehavior
{
    public ReferencePreservingDataContractSerializerOperationBehavior(OperationDescription operationDescription)
        : base(operationDescription) { }

    public override XmlObjectSerializer CreateSerializer(Type type, XmlDictionaryString name, XmlDictionaryString ns, IList<Type> knownTypes)
    {
        var settings = new DataContractSerializerSettings
        {
            RootName = name,
            RootNamespace = ns,
            KnownTypes = knownTypes,
            MaxItemsInObjectGraph = 0x7FFF,
            IgnoreExtensionDataObject = false,
            PreserveObjectReferences = true,
            DataContractResolver = null
        };

        return new DataContractSerializer(type, settings);
    }
}
```

**Reply by mconnew** [reply]
The code you have should work, it's the correct way to solve the problem. The only think I can think of that might be causing the problem is if you have the WCF Client packages added to your project as well as CoreWCF, are you implementing `CoreWCF.Description.IOperationBehavior` or `System.ServiceModel..Description.IOperationBehavior`?  

If that isn't the issue, can you create a small repro app and provide that so I can debug it?

**Reply by furqansafdar** [reply]
I tried both these options, first implementing through `System.ServiceModel..Description.IOperationBehavior` and later via `CoreWCF.Description.IOperationBehavior`. 

I have created this small [repo](https://github.com/furqansafdar/CoreWCF_Demo) to simulate the issue and still the error exists.

> System.ServiceModel.CommunicationException
>   HResult=0x80131500
>   Message=There was an error while trying to serialize parameter http://tempuri.org/:manager. The InnerException message was 'Object graph for type 'ClientDomain.Employee' contains cycles and cannot be serialized if references are not tracked. Consider using the DataContractAttribute with the IsReference property set to true.'.  Please see InnerException for more details.
>   Source=System.Private.ServiceModel
>   StackTrace:
>    at System.Runtime.AsyncResult.End[TAsyncResult](IAsyncResult result)
>    at System.ServiceModel.Channels.ServiceChannel.SendAsyncResult.End(SendAsyncResult result)
>    at System.ServiceModel.Channels.ServiceChannel.EndCall(String action, Object[] outs, IAsyncResult result)
>    at System.ServiceModel.Channels.ServiceChannelProxy.TaskCreator.<>c__DisplayClass1_0.<CreateGenericTask>b__0(IAsyncResult asyncResult)

Apart from this, I learned that we can even apply custom JSON Serialization and can handle cyclic dependency there as well. I would be keen to know the steps for implementing it if there are any advantages or performance gains associated.

**Reply by furqansafdar** [reply]
@mconnew awaiting for your feedback to help resolve this issue.

**Reply by alfonsovgs** [reply]
Hi @furqansafdar 👋

I have the same error, were you able to  resolve it?

I'm exploring some solutions like custom serialization, but I think core wcf can't support it.
